File: 2-add-two-numbers/add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, A: Optional[ListNode], B: Optional[ListNode]) -> Optional[ListNode]:
        dummy = curr = ListNode(-1)
        carry = 0
        while A or B:
            A_val, B_val = 0, 0
            if A:
                A_val = A.val
                A = A.next
            if B:
                B_val = B.val
                B = B.next
            combined = A_val + B_val + carry # 14
            carry = combined // 10 # 14 // 10 => 1
            curr.next = ListNode(combined % 10) # 14 % 10 => 4
            curr = curr.next
        if carry:
            curr.next = ListNode(carry)
        return dummy.next

File: 2-add-two-numbers/add-two-numbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
    const dummy = curr = new ListNode(-1, null)
    let carry = 0
    while (l1 || l2) {
        let total = carry
        if (l1) {
            total += l1.val
            l1 = l1.next
        }
        if (l2) {
            total += l2.val
            l2 = l2.next
        }
        carry = Math.floor(total / 10)
        curr.next = new ListNode(total % 10)
        curr = curr.next
    }
    if (carry) {
        curr.next = new ListNode(carry)
    }
    return dummy.next
};